addons/godot-next/nodes/callback_delegator.gd

```diff
@@ -0,0 +1,163 @@
+	if not _elements:
+		_elements = ResourceSet.new()
+	return [ PropertyInfo.new_resource("_elements").to_dict() ]
+
+func _ready() -> void:
+	_handle_notification("_ready")
+
+func _enter_tree() -> void:
+	var elements = _elements.get_data().values()
+	for element in elements:
+		if not element.get_owner():
+			_initialize_element(element)
+	_check_for_empty_callbacks()
+	
+	_handle_notification("_enter_tree")
+
+func _exit_tree() -> void:
+	_handle_notification("_enter_tree")
+	pass
+
+func _process(delta: float) -> void:
+	_handle_notification("_process", delta)
+
+func _physics_process(delta: float) -> void:
+	_handle_notification("_physics_process", delta)
+
+func _input(event: InputEvent) -> void:
+	_handle_notification("_input", event)
+
+func _unhandled_input(event: InputEvent) -> void:
+	_handle_notification("_unhandled_input", event)
+
+func _unhandled_key_input(event: InputEventKey) -> void:
+	_handle_notification("_unhandled_key_input", event)
+
+##### VIRTUALS #####
+
+##### PUBLIC METHODS #####
+
+func add_element(p_type: Script) -> Resource:
+	var elements = _elements.get_data()
+	
+	var ct = ClassType.new(p_type)
+	if not ct.is_type(_elements.get_type()):
+		return null
+	if has_element(p_type):
+		return get_element(p_type)
+		
+	var element: Resource = p_type.new()
+	var element_name: String = ct.get_script_class()
+	
+	elements[element_name] = element
+	_initialize_element(element)
+	
+	return element
+
+func get_element(p_type: Script) -> Resource:
+	var elements = _elements.get_data()
+	return elements.get(ClassType.new(p_type).get_script_class(), null)
+
+func has_element(p_type: Script) -> bool:
+	var elements = _elements.get_data()
+	return elements.has(ClassType.new(p_type).get_script_class())
+
+func remove_element(p_type: Script) -> bool:
+	var elements = _elements.get_data()
+	var element = get_element(p_type)
+	if element:
+		_remove_from_callbacks(element)
+		element.free()
+		return elements.erase(ClassType.new(p_type).get_script_class())
+	return false
+
+##### PRIVATE METHODS #####
+
+func _handle_notification(p_name: String, p_param = null) -> void:
+	if not p_param:
+		for a_element in _callbacks[p_name]:
+			a_element.call(p_name)
+	else:
+		for a_element in _callbacks[p_name]:
+			a_element.call(p_name, p_param)
+
+func _initialize_element(p_element: Resource) -> void:
+	if p_element.has_method("_awake"):
+		p_element.__awake(self)
+	#warning-ignore:return_value_discarded
+	p_element.connect("script_changed", self, "_refresh_callbacks", [p_element])
+	_add_to_callbacks(p_element)
+
+func _add_to_callbacks(p_element: Resource) -> void:
+	for callback in _callbacks:
+		if p_element.has_method(callback) and p_element.get_enabled():
+			_callbacks[callback][p_element] = null
+
+func _remove_from_callbacks(p_element: Resource) -> void:
+	for callback in _callbacks:
+		_callbacks[callback].erase(p_element)
+	_check_for_empty_callbacks()
+
+func _check_for_empty_callbacks() -> void:
+	for callback_key in _callbacks:
```

Comment:
for consistency's sake: `for a_callback in _callbacks: match a_callback:` Also more readable imo.